{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1ebcc361-ce5e-479a-94f4-b1cec138388e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5ceb24d3-ee81-44b6-a063-442e106cde53",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: scvi-tools in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (1.2.0)\n",
      "Requirement already satisfied: rich>=12.0.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (13.4.1)\n",
      "Requirement already satisfied: lightning>=2.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (2.4.0)\n",
      "Requirement already satisfied: numpyro>=0.12.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (0.15.3)\n",
      "Requirement already satisfied: pyro-ppl>=1.6.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (1.9.1)\n",
      "Requirement already satisfied: scipy in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (1.10.1)\n",
      "Requirement already satisfied: sparse>=0.14.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (0.15.4)\n",
      "Requirement already satisfied: docrep>=0.3.2 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (0.3.2)\n",
      "Requirement already satisfied: torchmetrics>=0.11.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (1.6.0)\n",
      "Requirement already satisfied: jaxlib in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (0.4.35)\n",
      "Requirement already satisfied: optax in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (0.2.4)\n",
      "Requirement already satisfied: tensorboard>=2.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (2.18.0)\n",
      "Requirement already satisfied: mudata>=0.1.2 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (0.3.1)\n",
      "Requirement already satisfied: scikit-learn>=0.21.2 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (1.2.2)\n",
      "Requirement already satisfied: xarray>=2023.2.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (2024.10.0)\n",
      "Requirement already satisfied: tqdm in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (4.65.0)\n",
      "Requirement already satisfied: torch in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (2.5.1)\n",
      "Requirement already satisfied: numpy<2.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (1.26.4)\n",
      "Requirement already satisfied: pandas in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (2.2.3)\n",
      "Requirement already satisfied: anndata in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (0.11.1)\n",
      "Requirement already satisfied: jax in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (0.4.35)\n",
      "Requirement already satisfied: ml-collections>=0.1.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (1.0.0)\n",
      "Requirement already satisfied: flax in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scvi-tools) (0.10.1)\n",
      "Requirement already satisfied: six in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from docrep>=0.3.2->scvi-tools) (1.16.0)\n",
      "Requirement already satisfied: PyYAML<8.0,>=5.4 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from lightning>=2.0->scvi-tools) (6.0)\n",
      "Requirement already satisfied: fsspec[http]<2026.0,>=2022.5.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from lightning>=2.0->scvi-tools) (2023.5.0)\n",
      "Requirement already satisfied: packaging<25.0,>=20.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from lightning>=2.0->scvi-tools) (24.1)\n",
      "Requirement already satisfied: typing-extensions<6.0,>=4.4.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from lightning>=2.0->scvi-tools) (4.12.2)\n",
      "Requirement already satisfied: pytorch-lightning in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from lightning>=2.0->scvi-tools) (2.4.0)\n",
      "Requirement already satisfied: lightning-utilities<2.0,>=0.10.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from lightning>=2.0->scvi-tools) (0.11.8)\n",
      "Requirement already satisfied: absl-py in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from ml-collections>=0.1.1->scvi-tools) (2.1.0)\n",
      "Requirement already satisfied: array-api-compat!=1.5,>1.4 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from anndata->scvi-tools) (1.9.1)\n",
      "Requirement already satisfied: h5py>=3.6 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from anndata->scvi-tools) (3.8.0)\n",
      "Requirement already satisfied: exceptiongroup in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from anndata->scvi-tools) (1.2.2)\n",
      "Requirement already satisfied: natsort in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from anndata->scvi-tools) (8.3.1)\n",
      "Requirement already satisfied: multipledispatch in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from numpyro>=0.12.1->scvi-tools) (0.6.0)\n",
      "Requirement already satisfied: ml-dtypes>=0.4.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from jax->scvi-tools) (0.5.0)\n",
      "Requirement already satisfied: opt-einsum in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from jax->scvi-tools) (3.4.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from pandas->scvi-tools) (2023.3)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from pandas->scvi-tools) (2023.3)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from pandas->scvi-tools) (2.9.0)\n",
      "Requirement already satisfied: pyro-api>=0.1.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from pyro-ppl>=1.6.0->scvi-tools) (0.1.2)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from rich>=12.0.0->scvi-tools) (2.18.0)\n",
      "Requirement already satisfied: markdown-it-py<3.0.0,>=2.2.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from rich>=12.0.0->scvi-tools) (2.2.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scikit-learn>=0.21.2->scvi-tools) (3.1.0)\n",
      "Requirement already satisfied: joblib>=1.1.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scikit-learn>=0.21.2->scvi-tools) (1.2.0)\n",
      "Requirement already satisfied: numba>=0.49 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from sparse>=0.14.0->scvi-tools) (0.60.0)\n",
      "Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from tensorboard>=2.0->scvi-tools) (0.7.2)\n",
      "Requirement already satisfied: grpcio>=1.48.2 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from tensorboard>=2.0->scvi-tools) (1.68.0)\n",
      "Requirement already satisfied: werkzeug>=1.0.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from tensorboard>=2.0->scvi-tools) (3.1.3)\n",
      "Requirement already satisfied: protobuf!=4.24.0,>=3.19.6 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from tensorboard>=2.0->scvi-tools) (5.28.3)\n",
      "Requirement already satisfied: setuptools>=41.0.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from tensorboard>=2.0->scvi-tools) (67.8.0)\n",
      "Requirement already satisfied: markdown>=2.6.8 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from tensorboard>=2.0->scvi-tools) (3.7)\n",
      "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.4.127 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (12.4.127)\n",
      "Requirement already satisfied: nvidia-cusolver-cu12==11.6.1.9 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (11.6.1.9)\n",
      "Requirement already satisfied: triton==3.1.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (3.1.0)\n",
      "Requirement already satisfied: nvidia-curand-cu12==10.3.5.147 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (10.3.5.147)\n",
      "Requirement already satisfied: nvidia-nvtx-cu12==12.4.127 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (12.4.127)\n",
      "Requirement already satisfied: nvidia-cublas-cu12==12.4.5.8 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (12.4.5.8)\n",
      "Requirement already satisfied: nvidia-cudnn-cu12==9.1.0.70 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (9.1.0.70)\n",
      "Requirement already satisfied: nvidia-cusparse-cu12==12.3.1.170 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (12.3.1.170)\n",
      "Requirement already satisfied: sympy==1.13.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (1.13.1)\n",
      "Requirement already satisfied: nvidia-nccl-cu12==2.21.5 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (2.21.5)\n",
      "Requirement already satisfied: filelock in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (3.12.0)\n",
      "Requirement already satisfied: nvidia-cufft-cu12==11.2.1.3 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (11.2.1.3)\n",
      "Requirement already satisfied: nvidia-nvjitlink-cu12==12.4.127 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (12.4.127)\n",
      "Requirement already satisfied: networkx in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (3.1)\n",
      "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.4.127 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (12.4.127)\n",
      "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.4.127 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (12.4.127)\n",
      "Requirement already satisfied: jinja2 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from torch->scvi-tools) (3.1.4)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from sympy==1.13.1->torch->scvi-tools) (1.3.0)\n",
      "Requirement already satisfied: orbax-checkpoint in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from flax->scvi-tools) (0.9.1)\n",
      "Requirement already satisfied: tensorstore in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from flax->scvi-tools) (0.1.68)\n",
      "Requirement already satisfied: msgpack in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from flax->scvi-tools) (1.0.5)\n",
      "Requirement already satisfied: chex>=0.1.87 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from optax->scvi-tools) (0.1.87)\n",
      "Requirement already satisfied: etils[epy] in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from optax->scvi-tools) (1.10.0)\n",
      "Requirement already satisfied: toolz>=0.9.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from chex>=0.1.87->optax->scvi-tools) (0.12.0)\n",
      "Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (3.8.4)\n",
      "Requirement already satisfied: requests in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (2.32.3)\n",
      "Requirement already satisfied: mdurl~=0.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from markdown-it-py<3.0.0,>=2.2.0->rich>=12.0.0->scvi-tools) (0.1.2)\n",
      "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from numba>=0.49->sparse>=0.14.0->scvi-tools) (0.43.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.1.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from werkzeug>=1.0.1->tensorboard>=2.0->scvi-tools) (2.1.5)\n",
      "Requirement already satisfied: nest_asyncio in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from orbax-checkpoint->flax->scvi-tools) (1.6.0)\n",
      "Requirement already satisfied: humanize in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from orbax-checkpoint->flax->scvi-tools) (4.11.0)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (4.0.2)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (6.0.4)\n",
      "Requirement already satisfied: charset-normalizer<4.0,>=2.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (3.3.2)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (1.9.2)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (24.2.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (1.3.1)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (1.3.3)\n",
      "Requirement already satisfied: importlib_resources in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from etils[epy]->optax->scvi-tools) (6.4.5)\n",
      "Requirement already satisfied: zipp in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from etils[epy]->optax->scvi-tools) (3.20.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from requests->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (3.10)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from requests->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (2024.8.30)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from requests->fsspec[http]<2026.0,>=2022.5.0->lightning>=2.0->scvi-tools) (2.2.3)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install scvi-tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ac8ae751-9703-4fc4-9dca-46a115099a73",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.11.1\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                 9.5.0\n",
      "asciitree           NA\n",
      "asttokens           NA\n",
      "cffi                1.17.1\n",
      "cloudpickle         2.2.1\n",
      "comm                0.2.2\n",
      "cycler              0.10.0\n",
      "cython_runtime      NA\n",
      "dask                2023.5.1\n",
      "dateutil            2.9.0\n",
      "debugpy             1.8.6\n",
      "decorator           5.1.1\n",
      "defusedxml          0.7.1\n",
      "entrypoints         0.4\n",
      "exceptiongroup      1.2.2\n",
      "executing           2.1.0\n",
      "fasteners           0.18\n",
      "h5py                3.8.0\n",
      "igraph              0.10.4\n",
      "ipykernel           6.26.0\n",
      "ipython_genutils    0.2.0\n",
      "jedi                0.19.1\n",
      "jinja2              3.1.4\n",
      "joblib              1.2.0\n",
      "kiwisolver          1.4.4\n",
      "leidenalg           0.9.1\n",
      "llvmlite            0.43.0\n",
      "markupsafe          2.1.5\n",
      "matplotlib          3.7.1\n",
      "matplotlib_inline   0.1.7\n",
      "mpl_toolkits        NA\n",
      "msgpack             1.0.5\n",
      "natsort             8.3.1\n",
      "numba               0.60.0\n",
      "numcodecs           0.11.0\n",
      "numpy               1.26.4\n",
      "packaging           24.1\n",
      "pandas              2.2.3\n",
      "parso               0.8.4\n",
      "pexpect             4.9.0\n",
      "pickleshare         0.7.5\n",
      "platformdirs        4.3.6\n",
      "prompt_toolkit      3.0.48\n",
      "psutil              6.0.0\n",
      "ptyprocess          0.7.0\n",
      "pure_eval           0.2.3\n",
      "pyarrow             12.0.0\n",
      "pydev_ipython       NA\n",
      "pydevconsole        NA\n",
      "pydevd              3.1.0\n",
      "pydevd_file_utils   NA\n",
      "pydevd_plugins      NA\n",
      "pydevd_tracing      NA\n",
      "pygments            2.18.0\n",
      "pyparsing           3.0.9\n",
      "pytz                2023.3\n",
      "scipy               1.10.1\n",
      "session_info        1.0.0\n",
      "setuptools          67.8.0\n",
      "six                 1.16.0\n",
      "sklearn             1.2.2\n",
      "sparse              0.15.4\n",
      "sphinxcontrib       NA\n",
      "stack_data          0.6.2\n",
      "tblib               1.7.0\n",
      "texttable           1.6.7\n",
      "threadpoolctl       3.1.0\n",
      "tlz                 0.12.0\n",
      "toolz               0.12.0\n",
      "torch               2.5.1+cu124\n",
      "torchgen            NA\n",
      "tornado             6.4.1\n",
      "tqdm                4.65.0\n",
      "traitlets           5.14.3\n",
      "typing_extensions   NA\n",
      "wcwidth             0.2.13\n",
      "yaml                6.0\n",
      "zarr                2.14.2\n",
      "zipp                NA\n",
      "zmq                 23.0.0\n",
      "zoneinfo            NA\n",
      "-----\n",
      "IPython             8.27.0\n",
      "jupyter_client      7.4.9\n",
      "jupyter_core        5.7.2\n",
      "notebook            6.5.7\n",
      "-----\n",
      "Python 3.10.11 (main, May 16 2023, 00:28:57) [GCC 11.2.0]\n",
      "Linux-5.10.0-33-amd64-x86_64-with-glibc2.39\n",
      "-----\n",
      "Session information updated at 2024-11-19 11:48\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 1\n",
    "sc.logging.print_versions()\n",
    "sc.settings.set_figure_params(dpi=80, frameon=False, figsize=(3, 3), facecolor=\"white\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a525c3cf-a195-426b-9e8b-a9c4d0877193",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting pyroe\n",
      "  Downloading pyroe-0.9.3-py3-none-any.whl (41 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m41.7/41.7 kB\u001b[0m \u001b[31m2.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: packaging>=21.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from pyroe) (24.1)\n",
      "Requirement already satisfied: scanpy>=1.8.2 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from pyroe) (1.9.3)\n",
      "Collecting pyranges==0.0.120\n",
      "  Downloading pyranges-0.0.120.tar.gz (687 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m687.9/687.9 kB\u001b[0m \u001b[31m12.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting pandas<2.0.0,>=1.3.0\n",
      "  Downloading pandas-1.5.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.1 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m12.1/12.1 MB\u001b[0m \u001b[31m70.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hCollecting biopython>=1.77\n",
      "  Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m71.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: cython in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from pyranges==0.0.120->pyroe) (3.0.2)\n",
      "Collecting ncls>=0.0.63\n",
      "  Downloading ncls-0.0.68-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.4 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.4/2.4 MB\u001b[0m \u001b[31m75.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: tabulate in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from pyranges==0.0.120->pyroe) (0.9.0)\n",
      "Collecting sorted_nearest>=0.0.33\n",
      "  Downloading sorted_nearest-0.0.39-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.0 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.0/6.0 MB\u001b[0m \u001b[31m24.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hCollecting pyrle\n",
      "  Downloading pyrle-0.0.40.tar.gz (460 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m460.4/460.4 kB\u001b[0m \u001b[31m35.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: natsort in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from pyranges==0.0.120->pyroe) (8.3.1)\n",
      "Requirement already satisfied: numpy in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from biopython>=1.77->pyroe) (1.26.4)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from pandas<2.0.0,>=1.3.0->pyroe) (2.9.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from pandas<2.0.0,>=1.3.0->pyroe) (2023.3)\n",
      "Requirement already satisfied: matplotlib>=3.4 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (3.7.1)\n",
      "Requirement already satisfied: joblib in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (1.2.0)\n",
      "Requirement already satisfied: tqdm in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (4.65.0)\n",
      "Requirement already satisfied: h5py>=3 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (3.8.0)\n",
      "Requirement already satisfied: scipy>=1.4 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (1.10.1)\n",
      "Requirement already satisfied: patsy in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (0.5.3)\n",
      "Requirement already satisfied: umap-learn>=0.3.10 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (0.5.3)\n",
      "Requirement already satisfied: session-info in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (1.0.0)\n",
      "Requirement already satisfied: networkx>=2.3 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (3.1)\n",
      "Requirement already satisfied: statsmodels>=0.10.0rc2 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (0.14.0)\n",
      "Requirement already satisfied: seaborn in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (0.12.2)\n",
      "Requirement already satisfied: numba>=0.41.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (0.60.0)\n",
      "Requirement already satisfied: scikit-learn>=0.22 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (1.2.2)\n",
      "Requirement already satisfied: anndata>=0.7.4 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scanpy>=1.8.2->pyroe) (0.11.1)\n",
      "Requirement already satisfied: array-api-compat!=1.5,>1.4 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from anndata>=0.7.4->scanpy>=1.8.2->pyroe) (1.9.1)\n",
      "Requirement already satisfied: exceptiongroup in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from anndata>=0.7.4->scanpy>=1.8.2->pyroe) (1.2.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from matplotlib>=3.4->scanpy>=1.8.2->pyroe) (0.11.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from matplotlib>=3.4->scanpy>=1.8.2->pyroe) (1.4.4)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from matplotlib>=3.4->scanpy>=1.8.2->pyroe) (4.39.4)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from matplotlib>=3.4->scanpy>=1.8.2->pyroe) (1.0.7)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from matplotlib>=3.4->scanpy>=1.8.2->pyroe) (9.5.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from matplotlib>=3.4->scanpy>=1.8.2->pyroe) (3.0.9)\n",
      "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from numba>=0.41.0->scanpy>=1.8.2->pyroe) (0.43.0)\n",
      "Requirement already satisfied: six>=1.5 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from python-dateutil>=2.8.1->pandas<2.0.0,>=1.3.0->pyroe) (1.16.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from scikit-learn>=0.22->scanpy>=1.8.2->pyroe) (3.1.0)\n",
      "Requirement already satisfied: pynndescent>=0.5 in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from umap-learn>=0.3.10->scanpy>=1.8.2->pyroe) (0.5.10)\n",
      "Requirement already satisfied: stdlib-list in /opt/conda/envs/spatialdata/lib/python3.10/site-packages (from session-info->scanpy>=1.8.2->pyroe) (0.8.0)\n",
      "Building wheels for collected packages: pyranges, pyrle\n",
      "  Building wheel for pyranges (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for pyranges: filename=pyranges-0.0.120-py3-none-any.whl size=697148 sha256=076f8b5f619c8992890a5e51c3b18b834b653b1f65e04b4e955a1fd9deef6e2e\n",
      "  Stored in directory: /home/jovyan/.cache/pip/wheels/50/42/88/e0b92772443143e000b8e21f7ef2a38286f8199ed2002c4726\n",
      "  Building wheel for pyrle (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for pyrle: filename=pyrle-0.0.40-cp310-cp310-linux_x86_64.whl size=763621 sha256=ecc04e0d41c27d3355f754679a2be5ca0294c1f760339ab95e5c0a358a7f141f\n",
      "  Stored in directory: /home/jovyan/.cache/pip/wheels/6f/63/81/037babd4f9a8998e7d8ebb7b110c3a5e870ffae2bc60ad8f6e\n",
      "Successfully built pyranges pyrle\n",
      "Installing collected packages: sorted_nearest, ncls, biopython, pandas, pyrle, pyranges, pyroe\n",
      "  Attempting uninstall: pandas\n",
      "    Found existing installation: pandas 2.2.3\n",
      "    Uninstalling pandas-2.2.3:\n",
      "      Successfully uninstalled pandas-2.2.3\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "xarray 2024.10.0 requires pandas>=2.1, but you have pandas 1.5.3 which is incompatible.\n",
      "xarray-spatial 0.3.7 requires numpy<=1.23.4, but you have numpy 1.26.4 which is incompatible.\n",
      "xarray-dataclasses 1.6.0 requires xarray<2023.0,>=2022.3; python_version >= \"3.9\" and python_version < \"3.12\", but you have xarray 2024.10.0 which is incompatible.\n",
      "squidpy 1.2.3 requires numba<0.56.0,>=0.55.0, but you have numba 0.60.0 which is incompatible.\n",
      "spatialdata 0.0.10 requires typing-extensions<4.6.0,>=4.0.0, but you have typing-extensions 4.12.2 which is incompatible.\n",
      "napari-spatialdata 0.2.4 requires typing-extensions<=4.5, but you have typing-extensions 4.12.2 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0mSuccessfully installed biopython-1.84 ncls-0.0.68 pandas-1.5.3 pyranges-0.0.120 pyrle-0.0.40 pyroe-0.9.3 sorted_nearest-0.0.39\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install pyroe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af3bf28-832c-4843-b64c-88e6da4a91dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Path to your Alevin output directory (which includes `quants_mat.gz`)\n",
    "adata = load_fry(\"/home/jovyan/ifbdata/spatial_cell_id/Kush/alignment/demo_multiscale_data_check/WS_SKN_KCL10525738_S1/alevin\", backend=\"alevin\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fbf5eaf-0fdd-4a11-be34-869c0612597d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "spatialdata",
   "language": "python",
   "name": "spatialdata"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
